### PR TITLE
Protocol Upgrade

### PR DESCRIPTION
**Pull request by samuelsadok**
Replace the user-oriented communication protocol with an improved machine oriented protocol.

The main differences are:
 - the device sends a JSON descriptor to announce it's parameters and functions to clients
 - CRC16 for UART
 - acknowledgements for parameter writes (and reads and function calls)

`test_communication.py` still provides the "p/v/c" commands but uses the new protocol internally. The tool now accepts a "--usb" and "--serial" parameter to select a specific device.

The old protocol is still operational on USB.

### CODE REVIEW

File: Firmware/MotorControl/crc.hpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+//  source: https://users.ece.cmu.edu/~koopman/crc/index.html
+// Also known as CRC-16-DNP
+constexpr uint16_t CRC16_DEFAULT = 0x3d65;
+
+// Calculates an arbitrary CRC for one byte.
+// Adapted from https://barrgroup.com/Embedded-Systems/How-To/CRC-Calculation-C-Code
+template<typename T, unsigned POLYNOMIAL>
+static T calc_crc(T remainder, uint8_t value) {
+    constexpr T BIT_WIDTH = (CHAR_BIT * sizeof(T));
+    constexpr T TOPBIT = (1 << (BIT_WIDTH - 1));
```
Comment (kalvdans) [inline Firmware/MotorControl/crc.hpp]: Signed overflow if BIT_WIDTH is 32 or 64. Use `1ull << (BIT_WIDTH - 1)` or `(T)1 << (BIT_WIDTH - 1)` instead.

**Comment by madcowswe** [comment]
Btw: This addresses issue #17.

File: Firmware/MotorControl/crc.hpp (hunk 1 above)
Comment (samuelsadok) [reply Firmware/MotorControl/crc.hpp]: Good catch, fixed. For T=int32_t I don't see any issue though.

Comment (madcowswe) [reply Firmware/MotorControl/crc.hpp]: I think in the int32_t case, the signed overflow is technically undefined (though in practice it's bit-wise equivalent to unsigned overflow, which is defined).

https://stackoverflow.com/a/16188846/1128910

**Review by kalvdans — APPROVED** [review]

**Comment by madcowswe** [comment]
Things to finish off:
- [x] Documentation on how to get started with the new ODrive Explore python
- [x] Clean out old communication Documentation (or move to depricated README page)
- [x] Note about how to switch to legacy protocol
- [x] Test that arduino library still works (Legacy on UART by default)

**Pull request merged**